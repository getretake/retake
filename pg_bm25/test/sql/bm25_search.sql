-- this is needed to ensure consistency of printouts with postgres versions older than 12. Can be
-- deleted if we drop support for postgres 11.
ALTER SYSTEM SET extra_float_digits TO 0;
select pg_reload_conf();


-- Basic search query
SELECT *
FROM bm25_search
WHERE bm25_search @@@ 'description:keyboard OR category:electronics OR rating>2';

-- With BM25 scoring
SELECT paradedb.rank_bm25(ctid), * 
FROM bm25_search 
WHERE bm25_search @@@ 'category:electronics OR description:keyboard';

-- Test JSON search 
SELECT *
FROM bm25_search
WHERE bm25_search @@@ 'metadata.color:white';

-- Test real-time search
INSERT INTO bm25_search (description, rating, category) VALUES ('New keyboard', 5, 'Electronics');
DELETE FROM bm25_search WHERE id = 1;
UPDATE bm25_search SET description = 'PVC Keyboard' WHERE id = 2;
SELECT *
FROM bm25_search
WHERE bm25_search @@@ 'description:keyboard OR category:electronics OR rating>2';

-- Test search in another namespace/schema
SELECT *
FROM paradedb.mock_items
WHERE mock_items @@@ 'description:keyboard';

-- Test search with default tokenizer: no results
SELECT *
FROM paradedb.mock_items
WHERE mock_items @@@ 'description:earbud';

