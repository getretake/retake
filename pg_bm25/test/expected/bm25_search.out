-- this is needed to ensure consistency of printouts with postgres versions older than 12. Can be
-- deleted if we drop support for postgres 11.
ALTER SYSTEM SET extra_float_digits TO 0;
select pg_reload_conf();
 pg_reload_conf 
----------------
 t
(1 row)

-- Basic search query
SELECT *
FROM bm25_search
WHERE bm25_search @@@ 'description:keyboard OR category:electronics OR rating>2';
 id |         description         | rating |  category   |  col_text   | col_varchar | col_smallint | col_bigint | col_integer | col_oid | col_float4 | col_float8 | col_numeric | col_decimal | col_real | col_double | col_bool |     col_json     |    col_jsonb     
----+-----------------------------+--------+-------------+-------------+-------------+--------------+------------+-------------+---------+------------+------------+-------------+-------------+----------+------------+----------+------------------+------------------
  2 | Plastic Keyboard            |      4 | Electronics | Sample text | Sample text |           10 |    1000000 |         100 |       1 |       10.5 |     100.55 |       99.99 |       88.88 |    77.77 |      66.66 | t        | {"key": "value"} | {"key": "value"}
  1 | Ergonomic metal keyboard    |      4 | Electronics | Sample text | Sample text |           10 |    1000000 |         100 |       1 |       10.5 |     100.55 |       99.99 |       88.88 |    77.77 |      66.66 | t        | {"key": "value"} | {"key": "value"}
 12 | Innovative wireless earbuds |      5 | Electronics | Sample text | Sample text |           10 |    1000000 |         100 |       1 |       10.5 |     100.55 |       99.99 |       88.88 |    77.77 |      66.66 | t        | {"key": "value"} | {"key": "value"}
 22 | Fast charging power bank    |      4 | Electronics | Sample text | Sample text |           10 |    1000000 |         100 |       1 |       10.5 |     100.55 |       99.99 |       88.88 |    77.77 |      66.66 | t        | {"key": "value"} | {"key": "value"}
 32 | Bluetooth-enabled speaker   |      3 | Electronics | Sample text | Sample text |           10 |    1000000 |         100 |       1 |       10.5 |     100.55 |       99.99 |       88.88 |    77.77 |      66.66 | t        | {"key": "value"} | {"key": "value"}
(5 rows)

-- With BM25 scoring
SELECT paradedb.rank_bm25(ctid), * 
FROM bm25_search 
WHERE bm25_search @@@ 'category:electronics OR description:keyboard';
 rank_bm25 | id |         description         | rating |  category   |  col_text   | col_varchar | col_smallint | col_bigint | col_integer | col_oid | col_float4 | col_float8 | col_numeric | col_decimal | col_real | col_double | col_bool |     col_json     |    col_jsonb     
-----------+----+-----------------------------+--------+-------------+-------------+-------------+--------------+------------+-------------+---------+------------+------------+-------------+-------------+----------+------------+----------+------------------+------------------
    5.3765 |  2 | Plastic Keyboard            |      4 | Electronics | Sample text | Sample text |           10 |    1000000 |         100 |       1 |       10.5 |     100.55 |       99.99 |       88.88 |    77.77 |      66.66 | t        | {"key": "value"} | {"key": "value"}
   4.93101 |  1 | Ergonomic metal keyboard    |      4 | Electronics | Sample text | Sample text |           10 |    1000000 |         100 |       1 |       10.5 |     100.55 |       99.99 |       88.88 |    77.77 |      66.66 | t        | {"key": "value"} | {"key": "value"}
   2.10964 | 12 | Innovative wireless earbuds |      5 | Electronics | Sample text | Sample text |           10 |    1000000 |         100 |       1 |       10.5 |     100.55 |       99.99 |       88.88 |    77.77 |      66.66 | t        | {"key": "value"} | {"key": "value"}
   2.10964 | 22 | Fast charging power bank    |      4 | Electronics | Sample text | Sample text |           10 |    1000000 |         100 |       1 |       10.5 |     100.55 |       99.99 |       88.88 |    77.77 |      66.66 | t        | {"key": "value"} | {"key": "value"}
   2.10964 | 32 | Bluetooth-enabled speaker   |      3 | Electronics | Sample text | Sample text |           10 |    1000000 |         100 |       1 |       10.5 |     100.55 |       99.99 |       88.88 |    77.77 |      66.66 | t        | {"key": "value"} | {"key": "value"}
(5 rows)

-- Test real-time search
INSERT INTO bm25_search (description, rating, category) VALUES ('New keyboard', 5, 'Electronics');
DELETE FROM bm25_search WHERE id = 1;
UPDATE bm25_search SET description = 'PVC Keyboard' WHERE id = 2;
SELECT *
FROM bm25_search
WHERE bm25_search @@@ 'description:keyboard OR category:electronics OR rating>2';
 id |         description         | rating |  category   |  col_text   | col_varchar | col_smallint | col_bigint | col_integer | col_oid | col_float4 | col_float8 | col_numeric | col_decimal | col_real | col_double | col_bool |     col_json     |    col_jsonb     
----+-----------------------------+--------+-------------+-------------+-------------+--------------+------------+-------------+---------+------------+------------+-------------+-------------+----------+------------+----------+------------------+------------------
    | New keyboard                |      5 | Electronics |             |             |              |            |             |         |            |            |             |             |          |            |          |                  | 
  2 | PVC Keyboard                |      4 | Electronics | Sample text | Sample text |           10 |    1000000 |         100 |       1 |       10.5 |     100.55 |       99.99 |       88.88 |    77.77 |      66.66 | t        | {"key": "value"} | {"key": "value"}
 12 | Innovative wireless earbuds |      5 | Electronics | Sample text | Sample text |           10 |    1000000 |         100 |       1 |       10.5 |     100.55 |       99.99 |       88.88 |    77.77 |      66.66 | t        | {"key": "value"} | {"key": "value"}
 22 | Fast charging power bank    |      4 | Electronics | Sample text | Sample text |           10 |    1000000 |         100 |       1 |       10.5 |     100.55 |       99.99 |       88.88 |    77.77 |      66.66 | t        | {"key": "value"} | {"key": "value"}
 32 | Bluetooth-enabled speaker   |      3 | Electronics | Sample text | Sample text |           10 |    1000000 |         100 |       1 |       10.5 |     100.55 |       99.99 |       88.88 |    77.77 |      66.66 | t        | {"key": "value"} | {"key": "value"}
(5 rows)

-- Test search in another namespace/schema
SELECT *
FROM paradedb.mock_items
WHERE mock_items @@@ 'description:keyboard';
 id |       description        | rating |  category   
----+--------------------------+--------+-------------
  2 | Plastic Keyboard         |      4 | Electronics
  1 | Ergonomic metal keyboard |      4 | Electronics
(2 rows)

-- Test search with stem tokenizer: this should have results even though no description contains exactly the word "earbud"
SELECT *
FROM paradedb.mock_items
WHERE mock_items @@@ 'description:earbud';
 id |         description         | rating |  category
----+-----------------------------+--------+-------------
 12 | Innovative wireless earbuds |      5 | Electronics
(1 row)
